src/shims/mod.rs

```diff
@@ -44,11 +52,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         // Try to see if we can do something about foreign items.
         if this.tcx.is_foreign_item(instance.def_id()) {
-            // An external function that we cannot find MIR for, but we can still run enough
+            let attrs = this.tcx.get_attrs(instance.def_id());
+            let link_name = match attr::first_attr_value_str_by_name(&attrs, sym::link_name) {
+                Some(name) => name.as_str(),
+                None => this.tcx.item_name(instance.def_id()).as_str(),
+            };
+            // Strip linker suffixes (seen on 32-bit macOS).
+            let link_name = link_name.trim_end_matches("$UNIX2003");
+
+            // An external function that we (possibly) cannot find MIR for, but we can still run enough
             // of them to make miri viable.
-            this.emulate_foreign_item(instance.def_id(), args, dest, ret)?;
-            // `goto_block` already handled.
```

Comment:
Please keep this comment.

Reply:
I've replaced it with a more accurate one.

Reply:
The old comment specifically mentioned `goto_block` because it is an important invariant that `find_fn` either returns a body or calls `goto_block`. So please make sure that this is still explicitly stated, like the old comment did.